python/google/protobuf/pyext/message.cc

```diff
@@ -1180,20 +1180,41 @@ int InitAttributes(CMessage* self, PyObject* args, PyObject* kwargs) {
       const FieldDescriptor* value_descriptor =
           descriptor->message_type()->FindFieldByName("value");
       if (value_descriptor->cpp_type() == FieldDescriptor::CPPTYPE_MESSAGE) {
-        Py_ssize_t map_pos = 0;
-        PyObject* map_key;
-        PyObject* map_value;
-        while (PyDict_Next(value, &map_pos, &map_key, &map_value)) {
-          ScopedPyObjectPtr function_return;
-          function_return.reset(PyObject_GetItem(map.get(), map_key));
-          if (function_return.get() == NULL) {
-            return -1;
+        if (PyDict_Check(value)) {
+          Py_ssize_t map_pos = 0;
+          PyObject* map_key;
+          PyObject* map_value;
+          while (PyDict_Next(value, &map_pos, &map_key, &map_value)) {
+            ScopedPyObjectPtr function_return;
+            function_return.reset(PyObject_GetItem(map.get(), map_key));
+            if (function_return.get() == NULL) {
+              return -1;
+            }
+            ScopedPyObjectPtr ok(PyObject_CallMethod(
+                function_return.get(), "MergeFrom", "O", map_value));
+            if (ok.get() == NULL) {
+              return -1;
+            }
           }
-          ScopedPyObjectPtr ok(PyObject_CallMethod(
-              function_return.get(), "MergeFrom", "O", map_value));
-          if (ok.get() == NULL) {
+        } else {
+          ScopedPyObjectPtr iter(PyObject_GetIter(value));
```

Comment:
Can PyDict reuse this code?

Reply:
I think so; was trying to avoid changing existing behaviour too much, but it should be basically equivalent.

I suppose I should just trust the tests :)